qpid-jms-client/src/main/java/org/apache/qpid/jms/transports/netty/NettyEventLoopGroupPoolFactory.java

```diff
@@ -0,0 +1,253 @@
+            }
+            return threads;
+         }
+
+         public boolean retain() {
+            while (true) {
+               final int currValue = refCnt.get();
+               if (currValue == 0) {
+                  // this has been already disposed!
+                  return false;
+               }
+               if (refCnt.compareAndSet(currValue, currValue + 1)) {
+                  return true;
+               }
+            }
+         }
+
+         @Override
+         public EventLoopGroup ref() {
+            if (refCnt.get() == 0) {
+               throw new IllegalStateException("the event loop group cannot be reused");
+            }
+            return group;
+         }
+
+         @Override
+         public void close() {
+            while (true) {
+               final int currValue = refCnt.get();
+               if (currValue == 0) {
+                  return;
+               }
+               if (refCnt.compareAndSet(currValue, currValue - 1)) {
+                  if (currValue == 1) {
+                     // if a racing thread has borrowed this lease, it would try already to set it to a new value:
+                     // this one is a best effort cleanup to help GC
+                     sharedEventLoopGroup.compareAndSet(this, null);
+                     Future<?> fut = group.shutdownGracefully(0, SHUTDOWN_TIMEOUT, TimeUnit.MILLISECONDS);
+                     if (!fut.awaitUninterruptibly(2 * SHUTDOWN_TIMEOUT)) {
+                        LOG.trace("Channel group shutdown failed to complete in allotted time");
+                     }
+                  }
+                  return;
+               }
+            }
+         }
+      }
+
+      private final AtomicReference<SharedGroupRef> sharedEventLoopGroup = new AtomicReference<>(null);
+
+      @Override
+      public Ref<EventLoopGroup> sharedGroupWith(final int threads) {
+         if (threads <= 0) {
+            throw new IllegalArgumentException("threads must be > 0");
+         }
+         while (true) {
+            SharedGroupRef sharedGroupRef = sharedEventLoopGroup.get();
+            if (sharedGroupRef != null && sharedGroupRef.retain()) {
+               validateSharedRef(sharedGroupRef, threads);
+               return new AtomicCloseableRef<>(sharedGroupRef);
+            }
+            synchronized (this) {
+               sharedGroupRef = sharedEventLoopGroup.get();
+               if (sharedGroupRef != null && sharedGroupRef.retain()) {
+                  validateSharedRef(sharedGroupRef, threads);
+                  return new AtomicCloseableRef<>(sharedGroupRef);
+               }
+               sharedGroupRef = new SharedGroupRef(createEventLoopGroup(threads, this, new QpidJMSThreadFactory("SharedNettyEventLoopGroup :(" + SHARED_NETTY_EVENT_LOOP_GROUP_SEQUENCE.incrementAndGet() + ")", true)), threads);
+               sharedEventLoopGroup.set(sharedGroupRef);
```

Comment:
The group here would be shared by different connections, so I cannot see how to make the information about it clearly visible, if not by modifying concurrently the thread factory each time a new connection use it (and doesn't seem a good option because can happen that 3 Netty thread(s) would handle hundreds of physical connections)

Reply:
Im not sure you fully read my comment. What I said was that this would seem to give the threads in the pool all the exact same name, which just seems silly. Giving threads their own name isnt unusual or complicated. Have the factory add a counter in this case or something, the same way its presumably done in all the existing thread pool implementations that tend to do something like that (e.g think 'pool-X-thread-Y' style naming). I clearly wouldnt want the factory to report _incorrect_ connection details in the thread name. 

I merely qualified that with the additional loss of being able to tell which connection the thread was operating for to illustrate the full extent of the change it would mean, going from having a very specific name based on the connection, to one that not only isnt connection-related, but actually has no separable name at all from the other threads in the pool. (Though it certainly would be possible to reinstate the name on a per-operation basis easily enough if desired, at least for the bits occurring outside Netty. Alternatively a bunch of logging will probably need to change to make it possible to discern the connection, otherwise it will likely be unsupportable).

Reply:
Sorry, I now understand what you mean: let me check what I can do here :+1: